channels/offline-channel-js/src/OfflineChannel.ts

```diff
@@ -0,0 +1,614 @@
+                _consecutiveErrors = 0;
+                _retryAt = null;
+               
+            };
+
+            _self.resume = () => {
+                _paused = false;
+                _clearScheduledTimer();
+                _setupInMemoTimer();
+                _setSendNextTimer();
+            };
+
+            
+            _self.onunloadFlush = () => {
+                if (!_paused) {
+                    //TODO: or should try send first
+                    let shouldContinue = true;
+                    while (_inMemoBatch.count() && shouldContinue) {
+                        shouldContinue = _flushInMemoItems(true);
+                    }
+                    // TODO: unloadprovider might send events out of order
+                }
+            };
+
+            _self.flush = (sync: boolean, callBack?: (flushComplete?: boolean) => void, sendReason?: SendRequestReason): boolean | void | IPromise<boolean> => {
+                // No op
+            };
+
+            _self.getOfflineListener = () => {
+                return _offlineListener;
+
+            };
+
+            _self._doTeardown = (unloadCtx?: IProcessTelemetryUnloadContext, unloadState?: ITelemetryUnloadState) => {
+                _self.onunloadFlush();
+                _offlineListener && runTargetUnload(_offlineListener, false);
+                let handler = _urlCfg.batchHandler;
+                handler && handler.teardown();
+                _clearScheduledTimer();
+                _initDefaults();
+            };
+
+
+            _self["_getDbgPlgTargets"] = () => {
+                return [_urlCfg, _inMemoBatch, _senderInst];
+            };
+            
+
+            function _initDefaults() {
+                _hasInitialized = false;
+                _paused = false;
+                _sender = new Sender();
+                _urlCfg = null;
+                _offlineListener = null;
+                _diagLogger = null;
+                _endpoint = null;
+                _inMemoBatch = null;
+                _convertUndefined = undefValue;
+                _maxBatchSize = null;
+                _sendNextBatchTimer = null;
+                _consecutiveErrors = null;
+                _retryAt = null;
+                _maxBatchInterval = null;
+                _senderInst = null;
+                _offineSupport = null;
+                _primaryChannelId = null;
+                _overrideIkey = null;
+            }
+
+
+            function _shouldCacheEvent(urlConfig: IUrlLocalStorageConfig, item: IPostTransmissionTelemetryItem) {
+                if ((item.persistence) < urlConfig.minPersistenceCacheLevel) {
+                    return false;
+                }
+
+                return true;
+            }
+
+
+            function _setupInMemoTimer() {
+                if (!_inMemoFlushTimer) {
+                    _inMemoFlushTimer = scheduleTimeout(() => {
+                        _flushInMemoItems();
+                        _inMemoFlushTimer.refresh();
```

Comment:
We should only "refresh" (restart) the timer if there are items waiting to be flushed.

I think (not finished reading all of this yet) that you would just need to remove this from here and add it as an else clause.

```ts
if (!_inMemoFlushTimer) {
    _inMemoFlushTimer = scheduleTimeout(() => {
        _flushInMemoItems();
    }, _inMemoTimerOut);
    _inMemoFlushTimer.unref();
} else {
    // Restart the timer if not already running https://nevware21.github.io/ts-utils/typedoc/interfaces/ITimerHandler.html#enabled
    _inMemoFlushTimer.enabled = true;
}                        
```

Reply:
Note: Teams are VERY sensitive (and have tests) which detect times that keep running for no reason, because it stops devices from going to (and staying) asleep, so any timers we create we need to make sure that they are not always "polling" (running) unless we have something that we are waiting for.